### PR TITLE
[SPARK-27355][SS] Make query execution more sensitive to epoch message late or lost

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

In [SPARK-23503](https://github.com/apache/spark/pull/20936), we enforce sequencing of committed epochs for Continuous Execution. In case a message for epoch n is lost and epoch (n + 1) is ready for commit before epoch n is, epoch (n + 1) will wait for epoch n to be committed first. With extreme condition, we will wait for `epochBacklogQueueSize` (10000 in default) epochs and then failed. There is no need to wait for such a long time before query fail **if there maybe some message LATE/LOST**. In this PR, we make the condition more sensitive.

## How was this patch tested?

update existing unit tests.

### CODE REVIEW

**Comment by uncleGen** [comment]
cc @jose-torres

**Comment by uncleGen** [comment]
pending on fixing unit test failure

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1453,6 +1453,13 @@ object SQLConf {
 
   val CONTINUOUS_STREAMING_EPOCH_BACKLOG_QUEUE_SIZE =
     buildConf("spark.sql.streaming.continuous.epochBacklogQueueSize")
+      .doc("The max number of entries to be stored in queue to wait for late epochs. " +
+        "If this parameter is exceeded by the size of the queue, stream will stop with an error.")
+      .intConf
+      .createWithDefault(10)
+
+  val CONTINUOUS_STREAMING_EPOCH_MSG_BACKLOG_QUEUE_SIZE =
+    buildConf("spark.sql.streaming.continuous.epochMessageBacklogQueueSize")
```
Comment (attilapiros) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: The doc of `CONTINUOUS_STREAMING_EPOCH_MSG_BACKLOG_QUEUE_SIZE` is a simple copy paste of the doc of `CONTINUOUS_STREAMING_EPOCH_BACKLOG_QUEUE_SIZE`, this does not seems to me right here!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala [hunk 2]
```diff
@@ -117,7 +117,7 @@ private[sql] object EpochCoordinatorRef extends Logging {
 private[continuous] class EpochCoordinator(
     writeSupport: StreamingWrite,
     stream: ContinuousStream,
-    query: ContinuousExecution,
```
Comment (attilapiros) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: I think the old naming was still right and this change just increasing the diff, as a `ContinuousExecution` is inherited from `StreamExecution` which is a `StreamingQuery`.

Or check for example `StreamingQueryWrapper` which has a member `_streamingQuery` with the type of `StreamExecution`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala [hunk 3]
```diff
@@ -224,24 +227,24 @@ private[continuous] class EpochCoordinator(
...
         resolveCommitsAtEpoch(epoch)
       }
       checkProcessingQueueBoundaries()
   }
 
   private def checkProcessingQueueBoundaries() = {
-    if (partitionOffsets.size > epochBacklogQueueSize) {
-      query.stopInNewThread(new IllegalStateException("Size of the partition offset queue has " +
-        "exceeded its maximum"))
+    if (partitionOffsets.size > epochMsgBacklogQueueSize) {
```
Comment (attilapiros) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: As I see you are reinterpret the new config `CONTINUOUS_STREAMING_EPOCH_MSG_BACKLOG_QUEUE_SIZE` in place of the old `CONTINUOUS_STREAMING_EPOCH_BACKLOG_QUEUE_SIZE `.

I think this is can be done (if good reasons supports it) as the old config was only introduced into Spark 3.0.0-SNAPSHOT in the PR https://github.com/apache/spark/pull/23156.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (uncleGen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: get it

**Comment by gaborgsomogyi** [comment]
As Attila pointed out it contains unnecessary changes which maybe distracted me but at the first glance I don't see the main value.

> There is no need to wait for such a long time before query fail if there maybe some message LATE/LOST.

Not sure if I understand the main reasoning. If one Kafka partition have some problem because that specific server is slower than other after 10 epoch just kill the query? If it's an intermittent problem not sure it's the right thing to do.

**Comment by uncleGen** [comment]
@gaborgsomogyi Thanks for your reply.  [#23156](https://github.com/apache/spark/pull/20936) introduced a maximum queue threshold before stop the stream with a error. In [#23156](https://github.com/apache/spark/pull/20936) , we used the same threshold for different queue, i.e. `partitionCommits`, `partitionOffsets` and `epochsWaitingToBeCommitted`. Generally, the size of `partitionCommits` and `partitionOffsets` grow much faster than `epochsWaitingToBeCommitted`. The stream may fail with 10 epochs if partition number is 100. However, we may wait for 10000 epochs before failure if partition number is 1 (if i understand correctly). Well, this is a harsh boundary condition. The main concern of PR is to split there two thresholds to make query execution more sensitive to epoch message late or lost. If you feel like `10 epoch` is too sensitive in some  intermittent problem, we can relax this condition to 100 or other.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala (hunk 2 above)
Comment (uncleGen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: reverted

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala (hunk 3 above)
Comment (uncleGen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: Hmm... You remind me, we'd better to use a new config even though `CONTINUOUS_STREAMING_EPOCH_BACKLOG_QUEUE_SIZE ` is a new introduced config in Spark 3.0.0-SNAPSHOT.

**Comment by uncleGen** [comment]
retest this please.

**Comment by gaborgsomogyi** [comment]
I see the main intention now. I agree that the different queues are filled up with different speed and considered this when the configuration added.

I think the threshold is configurable which makes the actual implementation flexible enough to handle this situation. From my point of view additional fine tuning doesn't really help but makes this code more complex which has to be maintained.

**Comment by uncleGen** [comment]
@gaborgsomogyi Maybe we can merge these two configs, and I prefer to keep the epoch threshold config. Then there is no need to check `partitionCommits` and `partitionOffsets` again. At one level, this makes sense. These queues would not be unbounded and consume up all the memory easily.

**Comment by uncleGen** [comment]
@gaborgsomogyi Could you please take a second glance?

**Comment by uncleGen** [comment]
cc @zsxwing

**Comment by uncleGen** [comment]
ping @gaborgsomogyi @zsxwing

**Comment by uncleGen** [comment]
@gaborgsomogyi I have unified the two configs. We will only check late epochs, but not check `epochBacklogQueueSize` again. As said above, checking late epochs is enough to avoid using too much memory and avoid waiting for long time before query fail if there maybe some messages LATE/LOST.

**Comment by uncleGen** [comment]
ping @gaborgsomogyi

**Comment by uncleGen** [comment]
retest this please

**Comment by uncleGen** [comment]
retest this please

**Pull request closed without merging**